File: scripts/convert-to-spark-format.mjs
#!/usr/bin/env node
/**
 * Konvertiert questions-az104.ts für GitHub Spark:
 * - Array.join("\n") wird zu Template Literals (Backticks)
 */

import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const inputFile = join(__dirname, '..', 'src', 'app', 'lab-az104-exam', 'questions-az104.ts');

console.log('Lese Datei:', inputFile);
let content = readFileSync(inputFile, 'utf-8');

let count = 0;

// Funktion zum Parsen eines Array-Literals
function parseArrayLiteral(arrayContent) {
  const lines = [];
  let currentToken = '';
  let inString = false;
  let stringChar = null;
  let bracketDepth = 0;
  let i = 0;
  
  while (i < arrayContent.length) {
    const char = arrayContent[i];
    const prevChar = i > 0 ? arrayContent[i - 1] : '';
    
    // Escape-Sequenzen handhaben
    if (char === '\\' && inString) {
      currentToken += char;
      i++;
      if (i < arrayContent.length) {
        currentToken += arrayContent[i];
      }
      i++;
      continue;
    }
    
    // String-Grenzen erkennen
    if ((char === '"' || char === "'") && prevChar !== '\\') {
      if (!inString) {
        inString = true;
        stringChar = char;
        currentToken += char;
      } else if (char === stringChar) {
        inString = false;
        currentToken += char;
        stringChar = null;
      } else {
        currentToken += char;
      }
      i++;
      continue;
    }
    
    if (inString) {
      currentToken += char;
      i++;
      continue;
    }
    
    // Nicht im String: Array-Verschachtelung tracken
    if (char === '[') {
      bracketDepth++;
      currentToken += char;
    } else if (char === ']') {
      bracketDepth--;
      currentToken += char;
    } else if (char === ',' && bracketDepth === 0) {
      // Element-Trenner auf oberster Ebene
      const trimmed = currentToken.trim();
      if (trimmed) {
        lines.push(trimmed);
      }
      currentToken = '';
    } else {
      currentToken += char;
    }
    
    i++;
  }
  
  // Letztes Element
  const trimmed = currentToken.trim();
  if (trimmed) {
    lines.push(trimmed);
  }
  
  return lines;
}

// Konvertiere alle [].join("\n") oder [].join('\n') zu Template Literals
// Verwende eine nicht-greedy Regex für besseres Matching
content = content.replace(
  /\[\s*\n([\s\S]*?)\]\.join\(["']\\n["']\)/g,
  (match, arrayContent) => {
    count++;
    
    // Parse Array-Elemente
    const elements = parseArrayLiteral(arrayContent);
    
    // Entferne Quotes von jedem Element
    const unquoted = elements.map(el => {
      let cleaned = el.trim();
      // Entferne äußere Quotes (aber behalte escaped Quotes)
      if ((cleaned.startsWith('"') && cleaned.endsWith('"')) ||
          (cleaned.startsWith("'") && cleaned.endsWith("'"))) {
        cleaned = cleaned.slice(1, -1);
      }
      return cleaned;
    });
    
    // Füge zusammen und escape für Template Literal
    const joined = unquoted.join('\\n');
    const escaped = joined
      .replace(/`/g, '\\`')
      .replace(/\$\{/g, '\\${');
    
    return '`' + escaped + '`';
  }
);

console.log(`✓ ${count} Array.join() Konstrukte konvertiert`);

// Schreibe zurück
writeFileSync(inputFile, content, 'utf-8');
console.log('✓ Datei erfolgreich aktualisiert!');

File: scripts/convert-to-spark-format.py
#!/usr/bin/env python3
"""
Konvertiert questions-az104.ts für GitHub Spark:
- Array.join("\\n") wird zu Template Literals (Backticks)
- Verbessert Lesbarkeit und Kompatibilität
"""

import re
import sys

def convert_array_join_to_template_literal(content):
    """Konvertiert [].join('\\n') Konstrukte zu Template Literals"""
    
    def replace_array(match):
        # Extrahiere den Array-Inhalt
        array_content = match.group(1)
        
        # Entferne führende/trailing Whitespace von jeder Zeile
        lines = []
        current_string = ""
        bracket_depth = 0
        in_string = False
        string_char = None
        i = 0
        
        # Parse den Array-Content
        while i < len(array_content):
            char = array_content[i]
            
            # String-Handling
            if char in ['"', "'"] and (i == 0 or array_content[i-1] != '\\\\'):
                if not in_string:
                    in_string = True
                    string_char = char
                elif char == string_char:
                    in_string = False
                    string_char = None
            
            # Wenn wir außerhalb eines Strings sind und ein Komma finden
            if not in_string and char == ',' and bracket_depth == 0:
                lines.append(current_string.strip())
                current_string = ""
            else:
                # Klammer-Tracking für verschachtelte Strukturen
                if not in_string:
                    if char == '[':
                        bracket_depth += 1
                    elif char == ']':
                        bracket_depth -= 1
                current_string += char
            
            i += 1
        
        # Letzten String hinzufügen
        if current_string.strip():
            lines.append(current_string.strip())
        
        # Verarbeite jede Zeile
        processed_lines = []
        for line in lines:
            # Entferne führende/trailing Quotes
            line = line.strip()
            if line.startswith('"') and line.endswith('"'):
                line = line[1:-1]
            elif line.startswith("'") and line.endswith("'"):
                line = line[1:-1]
            
            # Escape Backticks in der Zeile
            line = line.replace('`', '\\`')
            # Escape ${} in der Zeile
            line = re.sub(r'\$\{', r'\\${', line)
            
            processed_lines.append(line)
        
        # Erstelle Template Literal
        result = '`' + '\\n'.join(processed_lines) + '`'
        return result
    
    # Regex-Pattern für Array.join("\\n") oder Array.join('\\n')
    # Matche Arrays mit join
    pattern = r'\[\s*\n((?:[^[\]]|\[[^\]]*\])*?)\s*\]\.join\(["\']\\n["\']\)'
    
    # Ersetze alle Vorkommen
    converted = re.sub(pattern, replace_array, content, flags=re.MULTILINE | re.DOTALL)
    
    return converted

def main():
    input_file = r"c:\Users\mmuja\mainlab-ui\src\app\lab-az104-exam\questions-az104.ts"
    output_file = r"c:\Users\mmuja\mainlab-ui\src\app\lab-az104-exam\questions-az104.ts.new"
    
    print(f"Lese {input_file}...")
    with open(input_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    print("Konvertiere Array.join() zu Template Literals...")
    converted = convert_array_join_to_template_literal(content)
    
    print(f"Schreibe {output_file}...")
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(converted)
    
    print("✓ Konvertierung abgeschlossen!")
    print(f"Neue Datei: {output_file}")
    print("Bitte überprüfen und dann die alte Datei ersetzen.")

if __name__ == "__main__":
    main()
